### ISSUE TITLE
Should our dask requirement for pip be `dask[complete]`

### ISSUE THREAD

**Issue by vyasr** [rapidsai/rapids-dask-dependency#31]
Currently we list our pip dask requirement as simply `dask`, but that does not install components required to use e.g. dask.dataframe (such as pandas). What is the correct dependency to specify here? Should it be `dask[dataframe]`? `dask[complete]`? Something else? Do we need to make any changes to the conda dependency?

**Comment by vyasr** [comment]
CC @rjzamora @pentschev

**Comment by pentschev** [comment]
Neither of them are a requirement per se, so I'd say `dask` is "correct", anything else may be a requirement for specific libraries (e.g., `dask-cuda` may required `dask[dataframe]`, but it isn't a requirement for `dask-cuda`). With #27 are we still able to add patches (such as to `dask.dataframe`) if those aren't required and just patch if `dask.dataframe` is actually installed/loadable?

**Comment by vyasr** [comment]
Patches are only applied whenever a module is loaded, so if `dask.dataframe` is never loaded it's not an issue.

**Comment by vyasr** [comment]
> Neither of them are a requirement per se, so I'd say dask is "correct", anything else may be a requirement for specific libraries 

The issue with this is that if the definition of a dask extra changes (e.g. if dask[dataframe] adds a new dependency), the only way for a package using rapids-dask-dependency to match the right version of dask is to copy the same pinning. Concretely, imagine dask-cudf requires `rapids-dask-dependency` but also requires `dask[dataframe]`. The only way to ensure that this is consistent is for `dask-cudf` to also list `dask[dataframe]` with the same pinning of `dask` as is contained in `rapids-dask-dependency`, which then removes the benefit of centralizing the pinning.

I don't know exactly how this works in conda; does the dask package in conda just pull the dependencies associated with all pip extras (like dataframe)?

**Comment by pentschev** [comment]
> The only way to ensure that this is consistent is for `dask-cudf` to also list `dask[dataframe]` with the same pinning of `dask` as is contained in `rapids-dask-dependency`, which then removes the benefit of centralizing the pinning.

When you say "same pinning", do you mean it has to be exactly the same or must have an overlap? For example, if we now pin to `2024.1.1` and `dask-cudf` has `dask[dataframe]>=2024.1.0,<2024.2.0`, would that still be satisfied?

> I don't know exactly how this works in conda; does the dask package in conda just pull the dependencies associated with all pip extras (like dataframe)?

AFAICT, conda packages shouldn't be pulling pip extras, am I missing something here?

**Comment by vyasr** [comment]
> When you say "same pinning", do you mean it has to be exactly the same or must have an overlap? For example, if we now pin to 2024.1.1 and dask-cudf has dask[dataframe]>=2024.1.0,<2024.2.0, would that still be satisfied?

Yes, that will work. The case that I wasn't sure about is during nightly development when the dask dependency here is specified as just a git URL. I didn't think that will satisfy the "real" version pinning you show in dask-cudf.

However, from a quick experiment, it looks like at least modern versions of pip are smarter than I gave them credit for, so maybe we're OK here:

```
(rapids) coder _ ~/local/tmp $ pip download --no-deps 'dask[dataframe]>=2024.1.0,<2024.2.0' 'dask @ git+https://github.com/dask/dask.git@main'
Looking in indexes: https://pypi.org/simple, https://pypi.anaconda.org/rapidsai-wheels-nightly/simple
Collecting dask@ git+https://github.com/dask/dask.git@main
  Cloning https://github.com/dask/dask.git (to revision main) to /tmp/pip-download-6ts2gn30/dask_29e5f4daa43e4a408e3c2d599cdd56e3
  Running command git clone --filter=blob:none --quiet https://github.com/dask/dask.git /tmp/pip-download-6ts2gn30/dask_29e5f4daa43e4a408e3c2d599cdd56e3
  Resolved https://github.com/dask/dask.git to commit 5949e54bc04158d215814586a44d51e0eb4a964d
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Collecting dask<2024.2.0,>=2024.1.0 (from dask[dataframe]<2024.2.0,>=2024.1.0)
  Using cached dask-2024.1.1-py3-none-any.whl.metadata (3.7 kB)
ERROR: Cannot install dask 2024.3.0+1.g5949e54b (from git+https://github.com/dask/dask.git@main) and dask<2024.2.0 and >=2024.1.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested dask<2024.2.0 and >=2024.1.0
    The user requested dask 2024.3.0+1.g5949e54b (from git+https://github.com/dask/dask.git@main)

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

> AFAICT, conda packages shouldn't be pulling pip extras, am I missing something here?

Right, conda won't touch pip extras. What I more meant was, if you install dask from conda do you automatically get e.g. pandas? Or do you have to specify that separately?

It looks like the answer is that you get everything: [the conda recipe](https://github.com/conda-forge/dask-feedstock/blob/main/recipe/meta.yaml#L21) respecifies all the dependencies listed in [the pyproject optional-dependencies lists](https://github.com/dask/dask/blob/main/pyproject.toml#L47).

So I think we can close this issue as no action needed.

**Comment by jameslamb** [comment]
> *if you install dask from conda do you automatically get e.g. pandas?... looks like ... you get everything*

The `dask` conda package has all the dependencies of `dask[dataframe]` as you found.

Just want to add that there is also a `dask-core` package on conda-forge which has only the absolute minimum dependencies required to `import dask` ([conda-forge/dask-core-feedstock](https://github.com/conda-forge/dask-core-feedstock/blob/609e24586576fb614c3d5ca0ecdadd72fff950c6/recipe/meta.yaml#L26)).

Doesn't change the conclusion you reached here, but just wanted to mention it for completeness, and since `dask-core` might be the right choice in some conda situations in the future.

**Comment by pentschev** [comment]
@jameslamb is right. Given that, let me rephrase what I previously wrote in https://github.com/rapidsai/rapids-dask-dependency/issues/31#issuecomment-1981732693 :

> Neither of them are a requirement per se, so I'd say `dask` is "correct", anything else may be a requirement for specific libraries (e.g., `dask-cuda` may required `dask[dataframe]`, but it isn't a requirement for `dask-cuda`).

My comments about `dask-cuda` still apply, but this package is not about `dask-cuda` and rather about RAPIDS in general. With that said, we are currently adding both [`dask` and `dask-core`](https://github.com/rapidsai/rapids-dask-dependency/blob/3d6efa09d8955e276ab619ea850dede479d69238/conda/recipes/rapids-dask-dependency/meta.yaml#L18-L19) as dependencies to the conda-forge package, the latter is redundant, and _in the context of RAPIDS_ depending on `dask` is the correct choice. The same then could apply for wheels if we think that makes sense, and thus depending on `dask[dataframe]` is also correct, given RAPIDS is broadly dependent on dataframes.

**Comment by vyasr** [comment]
IIRC I added `dask-core` at someone's request at some point, but I no longer remember. I'm OK slimming down the dependencies however people see fit. The main tradeoff is that it'll be easier to maintain the union of all requirements here at the expense of bloating the dependency trees a bit for some packages that only need a subset of those dependencies.
